### PR TITLE
Vector Similarity - range query

### PR DESCRIPTION
**Pull request by alonre24**
VecSim range query feature based on the following design doc - https://redislabs.atlassian.net/wiki/spaces/DX/pages/3560538157/RediSearch+-+range+iterator+design

Main components: 
- Update parser to support vector range query syntax
- Add a new "Metric Iterator", which is a generic iterator that receives a list of doc ids and a list of numeric values. This iterator return results with/without the metric field.
- Use Metric iterator for vector query node of type RANGE, run the VecSim lib range query, and send the results to the iterator constructor.
- Extend and replace the metric yield mechanism - use a generic mechanism that allows an iterator to send a metric up the iterator's tree in every valid result. This includes an extension of the `RSIndexResult` with an array of `RLookup` key-value pairs that go up to the relevant RP, which in turn set the numeric value in the RLookup table, and returns it to the user.
- Apply this mechanism for metric iterator and hybrid iterator that uses for vector KNN query.

`test_vecsim.py` includes various tests and complex use-cases of the above.

**includes PRs**

- #3145 
- #3162 
- #3159 
- #3164 
- #3174

### CODE REVIEW

File: src/query_parser/v2/parser.y [hunk 1]
```diff
@@ -865,11 +860,9 @@ query ::= text_expr(A) ARROW LSQB vector_query(B) RSQB . { // main parse, hybrid
 
 query ::= star ARROW LSQB vector_query(B) RSQB . { // main parse, simple vecsim search as entire query case.
   setup_trace(ctx);
-  switch (B->vn.vq->type) {
-    case VECSIM_QT_KNN:
-      B->vn.vq->knn.order = BY_SCORE;
-      break;
-  }
+  RS_LOG_ASSERT(B->vn.vq->type == VECSIM_QT_KNN, "vector_query must be KNN");
+  B->vn.vq->knn.order = BY_SCORE;
```
Comment (ashtul) [inline src/query_parser/v2/parser.y:864]: add comment to explain why is different

File: src/query_parser/v2/parser.y [hunk 2]
```diff
@@ -955,11 +939,9 @@ query ::= text_expr(A) ARROW LSQB vector_query(B) RSQB ARROW LB attribute_list(C
 
 query ::= star ARROW LSQB vector_query(B) RSQB ARROW LB attribute_list(C) RB. {
   setup_trace(ctx);
-  switch (B->vn.vq->type) {
-    case VECSIM_QT_KNN:
-      B->vn.vq->knn.order = BY_SCORE;
-      break;
-  }
+  RS_LOG_ASSERT(B->vn.vq->type == VECSIM_QT_KNN, "vector_query must be KNN");
+  B->vn.vq->knn.order = BY_SCORE;
```
Comment (ashtul) [inline src/query_parser/v2/parser.y:943]: comment

File: src/redisearch.h [hunk 3]
```diff
@@ -250,24 +250,17 @@ typedef struct {
...
 typedef enum {
   RSResultType_Union = 0x1,
   RSResultType_Intersection = 0x2,
   RSResultType_Term = 0x4,
   RSResultType_Virtual = 0x8,
   RSResultType_Numeric = 0x10,
-  RSResultType_Distance = 0x20,
-  RSResultType_HybridDistance = 0x40,
+  RSResultType_Metric = 0x20,
+  RSResultType_HybridMetric = 0x40,
```
Comment (ashtul) [inline src/redisearch.h:260]: what is the difference?

File: src/result_processor.c [hunk 4]
```diff
@@ -245,61 +245,43 @@ ResultProcessor *RPScorer_New(const ExtScoringFunctionCtx *funcs,
...
-  for (size_t i = 0; i < self->nkeys; i++) {
-    RSValue *val;
-    if (res->indexResult->type == RSResultType_HybridDistance) {
-      val = RS_NumVal(res->indexResult->agg.children[0]->dist.distance);
-    } else {
-      // The entire query is a TOP-K query, or this is hybrid query that doesn't use the doc score,
-      // so the distance is saved in the root of indexResult.
-      val = RS_NumVal(res->indexResult->dist.distance);
-    }
-    RLookup_WriteOwnKey(self->keys[i], &(res->rowdata), val);
```
Comment (ashtul) [inline src/result_processor.c:283]: Why was this removed?

File: src/vector_index.c [hunk 5]
```diff
@@ -104,6 +131,37 @@ IndexIterator *NewVectorIterator(QueryEvalCtx *q, VectorQuery *vq, IndexIterator
       };
       return NewHybridVectorIterator(hParams);
     }
+    case VECSIM_QT_RANGE: {
```
Comment (ashtul) [inline src/vector_index.c:134]: IMO, this should be in a different function since there is no real correlation between `KNN` and `RANGE`.

It can be done later.

**Review by ashtul — COMMENTED** [review]
Looks ready.
Added a few questions.

File: src/query_parser/v2/parser.y (hunk 1 above)
Comment (alonre24) [reply src/query_parser/v2/parser.y:864]: When the query is "pure" KNN query (with no other filters) - we can return the results from vecsim lib ordered by score and avoid sorting by id (this is an optimization)

File: src/redisearch.h (hunk 3 above)
Comment (alonre24) [reply src/redisearch.h:260]: One is holding the metric value alone, the other is of type aggregate (used for hybrid queries), and has two children - 
1. `RSIndexResult` of type RSResultType_Metric that holds the metric.
2. `RSIndexResult` corresponding to the hybrid iterator's child's iterator type

File: src/result_processor.c (hunk 4 above)
Comment (alonre24) [reply src/result_processor.c:283]: This RP was replaced with a generic one - iterator that wants to export a metric write it directly to its `RSIndesResult`, and this RP is responsible for extracting the metric and insert it to the `RLookup` table (which is exactly what happened here before for the vecsim iterator)

**Review by ashtul — APPROVED** [review]
LGTM
